Submit Starter Code

lfu7 opened the pull request:
This pull request should address #6 , #20 .

- modified submission.ts to scrape different HTML elements conditioned on submission status
- modified helper.ts and interfaces.ts to accommodate new submission status
- created a submit() method in old_index.ts to get the results of a submission. returns a Submission object.
- created test cases in src/testcases for submit() method
- created a main() method in old_index.ts to test if submit() is working correctly

Modifications to consider:
- having a single Helper with one set of "session", "csrfToken" making HTTP requests has caused me to encounter "Rate Limit Exceeded" from Leetcode
   - solution: create User objects to make HTTP requests individually?
- error handling for in submission.detail(). Sometimes the response could be missing fields that we are scraping for. I think it   would make sense to wrap the code with try...except blocks

grifkin commented:
Looks good. I've been testing it and I it seems to work well. One thing I think should be added is the total_correct and total_testcases attributes of submission. These could be passed to the front end so the player could know how many test cases each of his submissions passed.

To get this working with sockets, I think the best approach would be to integrate these functions into the Race class in src/socket/race.ts. A socket event listener could listen for a "submit" event that passes the code and language used. This listener could then call some function in Race that handles the submission and emits a socket event to the Race's socket room with details of the results.

grifkin commented:
I'll leave this open for now since I didn't have time to test it too thoroughly. I'll be able to look over it more tonight.

lfu7 commented:
total_correct and total_testcases should already be attributes of submission.

I've changed the code from the last sprint. Getting a problem by difficulty is no longer a route. Now all of that logic should be in the static class src/lib/LeetcodeProblems.

I've added some code in race.ts to get a problem.

grifkin commented on src/socket/race.ts:
```diff
@@ -1,14 +1,29 @@
+	constructor(roomKey: string, difficulty: number) {
 		this.roomKey = roomKey;
 		console.log("Race object " + roomKey + " created with difficulty: " + difficulty);
 		// TODO: fetch problem based on difficulty
 		// and set this.problem to point to it
+		Race.setProblem(difficulty);
+	}
+
+	static async setProblem(difficulty: number) {
```
I think this could cause issues, since the constructor returns an object immediately, so the Race's problem might be null when other code interacts with it. 

A better way to do this might be to leave the problem null in the constructor, and instead make an await call to setProblem in the Tracker's start method.

	// start a race
	async start(code: string): boolean {
		if(!this.openLobbies.has(code)) {
			console.log("invalid code");
			return false;
		}

		// create a new race, pass in room code and difficulty
		let race: Race = new Race(code, this.openLobbies.get(code));
            await race.setProblem(this.openLobbies.get(code));
		this.activeRaces.add(race);
		this.removeLobby(code);
		return true;
	}

Maybe something like this. This way the socket route could await this start call before sending the "start" method to the socket room.

grifkin replied:
Also I don't think this method should be static, since static methods operate on a class while normally methods operate per object. Here we want the method to be object-specific.

lfu7 commented:
Should have just changed the setProblem() method to be non-static, will work the sockets later today.

Nopekick commented on src/socket/route.ts:
```diff
@@ -52,12 +56,68 @@ io.of("/race/private").use(function(socket, next) {
+		const problem = await tracker.start(code);
+
+		if(problem != null) {
+			const problem_details:JSON = <JSON><unknown>{
```
A better way to do this would be to have the object problem_details as 'const problem_details = {...}', and then do 
'io.of("/race/private").to(code).emit("start", JSON.stringify(problem_details));'

Nopekick commented on src/socket/route.ts:
```diff
@@ -52,12 +56,68 @@ io.of("/race/private").use(function(socket, next) {
+        		"input" : submission.input,
+        		"expected_output": submission.expected_output,
+			  }
+			io.of("/race/private").to(code).emit("submit", submission_details);
```
I'm assuming this is the notification to all players of the recent submission. This notification only needs to contain total_testcases and total_correct. Also, the name 'notification' would be more fitting. You can do something like 'const notification = {total: submission.total_testcases, passed: submission.total_correct}; io.of(...).to(code).emit("notification", JSON.stringify(notification));

Then, in addition to the notification, the person who just submitted their code needs to see all of the submission detail. Add 'io.to(socket).emit(JSON.stringify(submission_details))', as I believe that sends the data only to that person.

Nopekick reviewed (changes requested):
See the comments.

Nopekick commented on src/lib/submission.ts:
```diff
@@ -1,3 +1,4 @@
+import e from "cors";
```
What is this used for?

Nopekick commented on src/socket/route.ts:
```diff
@@ -14,8 +14,8 @@ let tracker = new Tracker();
 io.of("/race/private").use(function(socket, next) {
-	if(socket.handshake.query && socket.handshake.query.token) {
-		jwt.verify(socket.handshake.query.token, privateKey, function(err, decoded) {
+	if(socket.handshake.auth && socket.handshake.auth.token) {
```
Why was this changed back? The frontend works with auth

grifkin replied:
Postman doesn't support auth, so for testing we had to temporarily change it back to query
